django/db/models/query.py

```diff
@@ -608,7 +608,12 @@ def update_or_create(self, defaults=None, **kwargs):
                 return obj, created
             for k, v in resolve_callables(defaults):
                 setattr(obj, k, v)
-            obj.save(using=self.db)
+            # `update_fields` does not support non-concrete fields yet (#31382)
+            fnames = map(lambda f: f.name, self.model._meta.concrete_fields)
```

Comment:
You'd need to either include both `f.name` and `f.attname` or use `self.model._meta.get_field(name)` for each `defaults` which I think supports both form

e.g.

```python
get_field = self.model._meta.get_field
update_defaults = True
for default in defaults:
    try:
        field = get_field(default)
    except FieldError:
        break
    if not field.concrete:
        break
else:
    update_defaults = False
```

Reply:
Yeah, I copied the code from `_save_table` in the meantime -- not sure which variant is the best :) And yes it feels wasteful, but what are the options? I mean the best thing to do would be to fix update_fields to operate properly on non-concrete fields…

Reply:
> And yes it feels wasteful, but what are the options?

I guess `Options` could have a private `cached_property(_concrete_field_names -> frozenset)`. It seems niche enough to be kept private but worth it given this set is computed on every `Model.save(update_fields)` call and on every `QuerySet.update_or_create` call after this PR.